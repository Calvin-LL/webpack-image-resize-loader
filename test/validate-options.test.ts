import webpack from "webpack";

import compile from "./helpers/compile";
import getCompiler from "./helpers/getCompiler";

describe("validate options", () => {
  const tests = {
    size: {
      success: [
        {
          width: 10,
        },
      ],
      failure: [true],
    },
    scale: {
      success: [0.1, 0.5, 1],
      failure: [-1, 0, 1.1, 2],
    },
    format: {
      success: ["jpeg", "png", "webp", "tiff"],
      failure: [true],
    },
    quality: {
      success: [10],
      failure: [0],
    },
    scaleUp: {
      success: [true, false],
      failure: [0],
    },
    sharpOptions: {
      success: [{ resize: { fastShrinkOnLoad: true } }],
      failure: [0],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = getCompiler({
        size: {
          width: 10,
        },
        ...{ [key]: value },
      });

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          expect((stats as webpack.Stats).hasErrors()).toBe(false);
        } else if (type === "failure") {
          const {
            compilation: { errors },
          } = stats as any;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
