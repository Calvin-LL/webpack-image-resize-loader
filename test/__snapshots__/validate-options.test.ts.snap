// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "format" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.format should be one of these:
   \\"jpeg\\" | \\"png\\" | \\"webp\\" | \\"tiff\\"
   -> The format of the output file. When unspecified, outputs the same format as the imported file."
`;

exports[`validate options should throw an error on the "quality" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.quality should be >= 1.
   -> The quality of the output image. Default to 80 for JPEG, WebP, and TIFF, default to 100 for PNG."
`;

exports[`validate options should throw an error on the "scaleUp" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scaleUp should be a boolean.
   -> Whether or not to scale up the image when the desired size is larger than the image size. Default to false."
`;

exports[`validate options should throw an error on the "sharpOptions" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.sharpOptions should be an object:
   object { resize?, png?, jpeg?, webp?, tiff?, … }"
`;

exports[`validate options should throw an error on the "size" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.size should be an object:
   object { width?, height?, fit?, position?, background?, … }
   -> The size of the output image. This field is required."
`;

exports[`validate options should throw an error on the "size" option with undefined value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options misses the property 'size'. Should be:
   object { width?, height?, fit?, position?, background?, … }
   -> The size of the output image. This field is required."
`;
